src/ipvs/ip_vs_proto_tcp.c

```diff
@@ -306,46 +306,88 @@ static void tcp_in_remove_ts(struct tcphdr *tcph)
     }
 }
 
-/* use NOP option to replace TCP_OLEN_IP4_ADDR and TCP_OLEN_IP6_ADDR opt */
-static void tcp_in_remove_toa(struct tcphdr *tcph, int af)
+/*
+ * Remove NOP and TOA options preset in the mbuf and compact option space.
+ * Return the trimmed length on success, otherwise dpvs error num on failure.
+ * */
+static int tcp_in_prune_options(int af, struct rte_mbuf *mbuf, struct tcphdr *tcph)
 {
-    unsigned char *ptr;
-    int len, i;
-    uint32_t tcp_opt_len = af == AF_INET ? TCP_OLEN_IP4_ADDR : TCP_OLEN_IP6_ADDR;
+    unsigned char *ptr, *fast, *slow;
+    const unsigned char *l3hdr, *payload;
+    int i, optlen;
+    unsigned int pruned;
+    uint8_t opcode, opsize;
 
     ptr = (unsigned char *)(tcph + 1);
-    len = (tcph->doff << 2) - sizeof(struct tcphdr);
-
-    while (len > 0) {
-        int opcode = *ptr++;
-        int opsize;
+    fast = slow = ptr;
+    optlen = (tcph->doff << 2) - sizeof(struct tcphdr);
+    payload = ptr + optlen;
 
+    while (optlen > 0) {
+        opcode = *ptr++;
         switch (opcode) {
         case TCP_OPT_EOL:
-            return;
+            goto fini;
```

Comment:
set `fast += optlen` to trim unused options?

Reply:
`payload` , a pointer to the beginning of l4 payload, is used instead of the `fast` pointer when trimming the packet. I think it's an equivalent to what you mean, and in a safer way.